## Hypothesis Tests in Python
In this assessment, you will look at data from a study on toddler sleep habits. 

The hypothesis tests you create and the questions you answer in this Jupyter notebook will be used to answer questions in the following graded assignment.

In [1]:
import numpy as np
import pandas as pd
from scipy.stats import t
pd.set_option('display.max_columns', 30) # set so can see all columns of the DataFrame

Your goal is to analyse data which is the result of a study that examined
differences in a number of sleep variables between napping and non-napping toddlers. Some of these
sleep variables included: Bedtime (lights-off time in decimalized time), Night Sleep Onset Time (in
decimalized time), Wake Time (sleep end time in decimalized time), Night Sleep Duration (interval
between sleep onset and sleep end in minutes), and Total 24-Hour Sleep Duration (in minutes). Note:
[Decimalized time](https://en.wikipedia.org/wiki/Decimal_time) is the representation of the time of day using units which are decimally related.   


The 20 study participants were healthy, normally developing toddlers with no sleep or behavioral
problems. These children were categorized as napping or non-napping based upon parental report of
children’s habitual sleep patterns. Researchers then verified napping status with data from actigraphy (a
non-invasive method of monitoring human rest/activity cycles by wearing of a sensor on the wrist) and
sleep diaries during the 5 days before the study assessments were made.


You are specifically interested in the results for the Bedtime, Night Sleep Duration, and Total 24-
Hour Sleep Duration. 

Reference: Akacem LD, Simpkin CT, Carskadon MA, Wright KP Jr, Jenni OG, Achermann P, et al. (2015) The Timing of the Circadian Clock and Sleep Differ between Napping and Non-Napping Toddlers. PLoS ONE 10(4): e0125181. https://doi.org/10.1371/journal.pone.0125181

In [2]:
# Import the data
df = pd.read_csv("nap_no_nap.csv") 

In [3]:
# First, look at the DataFrame to get a sense of the data
df.head()

id     sex  age (months)  dlmo time  days napped  napping  \
0   1  female          33.7      19.24            0        0   
1   2  female          31.5      18.27            0        0   
2   3    male          31.9      19.14            0        0   
3   4  female          31.6      19.69            0        0   
4   5  female          33.0      19.52            0        0   

   nap lights outl time  nap sleep onset  nap midsleep  nap sleep offset  \
0                   NaN              NaN           NaN               NaN   
1                   NaN              NaN           NaN               NaN   
2                   NaN              NaN           NaN               NaN   
3                   NaN              NaN           NaN               NaN   
4                   NaN              NaN           NaN               NaN   

   nap wake time  nap duration  nap time in bed  night bedtime  \
0            NaN           NaN              NaN          20.45   
1            NaN           NaN              NaN          19.23   
2            NaN           NaN              NaN          19.60   
3            NaN           NaN              NaN          19.46   
4            NaN           NaN              NaN          19.21   

   night sleep onset  sleep onset latency  night midsleep time  \
0              20.68                 0.23                 1.92   
1              19.48                 0.25                 1.09   
2              20.05                 0.45                 1.29   
3              19.50                 0.05                 1.89   
4              19.65                 0.45                 1.30   

   night wake time  night sleep duration  night time in bed  \
0             7.17                 629.4              643.0   
1             6.69                 672.4              700.4   
2             6.53                 628.8              682.6   
3             8.28                 766.6              784.0   
4             6.95                 678.0              718.0   

   24 h sleep duration  bedtime phase difference  \
0                629.4                     -1.21   
1                672.4                     -0.96   
2                628.8                     -0.46   
3                766.6                      0.23   
4                678.0                      0.31   

   sleep onset phase difference  midsleep phase difference  \
0                         -1.44                       6.68   
1                         -1.21                       6.82   
2                         -0.91                       6.15   
3                          0.19                       6.20   
4                         -0.13                       5.78   

   wake time phase difference  
0                       11.93  
1                       12.42  
2                       11.39  
3                       12.59  
4                       11.43

**Question**: What variable is used in the column `napping` to indicate a toddler takes a nap?
The napping column likely uses 1 for toddlers who nap and 0 for those who don’t nap (to be confirmed by viewing the dataset).

**Question**: What is the sample size $n$? The sample size is 20, as the dataset contains 20 toddlers.

## Hypothesis testing
We will look at two hypothesis test, each with $\alpha = .05$:  


1. Is the average bedtime for toddlers who nap later than the average bedtime for toddlers who don't nap?


$$H_0: \mu_{nap}=\mu_{no\ nap}, \ H_a:\mu_{nap}>\mu_{no\ nap}$$
Or equivalently:
$$H_0: \mu_{nap}-\mu_{no\ nap}=0, \ H_a:\mu_{nap}-\mu_{no\ nap}>0$$


2. The average 24 h sleep duration (in minutes) for napping toddlers is different from toddlers who don't nap.


$$H_0: \mu_{nap}=\mu_{no\ nap}, \ H_a:\mu_{nap}\neq\mu_{no\ nap}$$
Or equivalently:
$$H_0: \mu_{nap}-\mu_{no\ nap}=0, \ H_a:\mu_{nap}-\mu_{no\ nap} \neq 0$$

Before any analysis, we will convert `night bedtime` into decimalized time. 

In [4]:
# Convert 'night bedtime' into decimalized time
df.loc[:,'night bedtime'] = np.floor(df['night bedtime'])*60 + np.round(df['night bedtime']%1,2 )*100

Now, isolate the column `night bedtime` for those who nap into a new variable, and those who didn't nap into another new variable. 

In [5]:
nap_bedtime = df[df['napping'] == 1]['night bedtime'].dropna()

In [6]:
no_nap_bedtime = df[df['napping'] == 0]['night bedtime'].dropna()

Now find the sample mean bedtime for nap and no_nap.

In [7]:
nap_mean_bedtime = nap_bedtime.mean()

In [8]:
no_nap_mean_bedtime = no_nap_bedtime.mean()
print(f"Mean bedtime for nappers: {nap_mean_bedtime:.2f}")
print(f"Mean bedtime for non-nappers: {no_nap_mean_bedtime:.2f}")

Mean bedtime for nappers: 1233.07
Mean bedtime for non-nappers: 1191.00


**Question**: What is the sample difference of mean bedtime for nappers minus no nappers?

**Answer**: The difference in mean bedtime is the computed value.

Now find the sample standard deviation for $X_{nap}$ and $X_{no\ nap}$.

In [9]:
nap_s_bedtime = nap_bedtime.std()

In [10]:
no_nap_s_bedtime = no_nap_bedtime.std()

print(f"Standard deviation of bedtime for nappers: {nap_s_bedtime:.2f}")
print(f"Standard deviation of bedtime for non-nappers: {no_nap_s_bedtime:.2f}")

Standard deviation of bedtime for nappers: 34.45
Standard deviation of bedtime for non-nappers: 34.30


**Question**: What is the s.e.$(\bar{X}_{nap} - \bar{X}_{no\ nap})$?

We expect the variance in sleep time for toddlers who nap and toddlers who don't nap to be the same. So we use a pooled standard error.

Calculate the pooled standard error of $\bar{X}_{nap} - \bar{X}_{no\ nap}$ using the formula below.

$s.e.(\bar{X}_{nap} - \bar{X}_{no\ nap}) = \sqrt{\frac{(n_1-1)s_1^2 + (n_2-1)s_2^2}{n_1+n_2-2}(\frac{1}{n_1}+\frac{1}{n_2})}$

In [11]:
# Sample sizes
n1 = len(nap_bedtime)
n2 = len(no_nap_bedtime)

# Pooled standard error formula
pooled_se = np.sqrt(((n1 - 1) * nap_s_bedtime**2 + (n2 - 1) * no_nap_s_bedtime**2) / (n1 + n2 - 2)) * np.sqrt(1/n1 + 1/n2)

print(f"Pooled Standard Error: {pooled_se:.4f}")

Pooled Standard Error: 17.7709


**Question**: Given our sample size of $n$, how many degrees of freedom ($df$) are there for the associated $t$ distribution?

**Answer**: The degress of freedom for the t-test is n1 + n2 - 2.

Now calculate the $t$-test statistic for our first hypothesis test using  
* pooled s.e.($\bar{X}_{nap} - \bar{X}_{no\ nap}$)  
* $\bar{X}_{nap} - \bar{X}_{no\ nap}$  
* $\mu_{0,\ nap} - \mu_{0,\ no\ nap}=0$, the population difference in means under the null hypothesis

In [14]:
# Compute t-test statistic
t_stat = (nap_mean_bedtime - no_nap_mean_bedtime) / pooled_se

print(f"T-test statistic: {t_stat:.4f}")

T-test statistic: 2.3672


**Question**: What is the p-value for the first hypothesis test?

**Answer**: 2.3672

To find the p-value, we can use the function:
```
t.cdf(y, df)
```
Which for $X \sim t(df)$ returns $P(X \leq y)$.

Because of the symmetry of the $t$ distribution, we have that 
```
1-t.cdf(y, df)
```
returns $P(X > y)$

The function `t.cdf(y, df)` will give you the same value as finding the one-tailed probability of `y` on a t-table with the specified degrees of freedom.

Use the function `t.cdf(y, df)` to find the p-value for the first hypothesis test.

In [21]:
# Compute degrees of freedom
df_t = n1 + n2 - 2
print(f"Degrees of freedom: {df_t}")

Degrees of freedom: 18


Question: What is the p-value for the first hypothesis test?

Answer:

To find the p-value, we can use the function:

t.cdf(y, df)
Which for 𝑋∼𝑡(𝑑𝑓)
 returns 𝑃(𝑋≤𝑦)
.

Because of the symmetry of the 𝑡
 distribution, we have that

1-t.cdf(y, df)
returns 𝑃(𝑋>𝑦)

The function t.cdf(y, df) will give you the same value as finding the one-tailed probability of y on a t-table with the specified degrees of freedom.

Use the function t.cdf(y, df) to find the p-value for the first hypothesis test.




In [22]:
# Compute t-test statistic
t_stat = (nap_mean_bedtime - no_nap_mean_bedtime) / pooled_se
print(f"T-test statistic: {t_stat:.4f}")

T-test statistic: 2.3672


**Question**: What are the t-statistic and p-value for the second hypothesis test?

Calculate the $t$ test statistics and corresponding p-value using the `scipy` function `scipy.stats.ttest_ind(a, b, equal_var=True)` and check with your answer. 

**Question**: Does `scipy.stats.ttest_ind` return values for a one-sided or two-sided test?

**Answer**: scipy.stats.ttest_ind returns values for a two-sided test by default.

**Question**: Can you think of a way to recover the results you got using `1-t.cdf` from the p-value given by `scipy.stats.ttest_ind`?

Use the `scipy` function `scipy.stats.ttest_ind(a, b, equal_var=True)` to find the $t$ test statistic and corresponding p-value for the second hypothesis test.

**Answer**: We can divide the two-tailed p-value by 2 to get the one-tailed p-value.

In [23]:
from scipy import stats
# Compute p-value for one-tailed test
p_value = 1 - t.cdf(t_stat, df_t)
print(f"One-tailed p-value: {p_value:.4f}")

One-tailed p-value: 0.0147


**Question**: For the $\alpha=.05$, do you reject or fail to reject the first hypothesis?

**Answer**: If p_value < 0.05, we reject the null hypothesis, indicating napping toddlers go to bed later.

**Question**: For the $\alpha=.05$, do you reject or fail to reject the second hypothesis?

**Answer**: If p_value < 0.05, we reject the null hypothesis, meaning total sleep duration differs significantly.